flight/Modules/ManualControl/manualcontrol.c

```diff
@@ -136,112 +143,195 @@ static void manualControlTask(void *parameters)
-				control_events = tablet_control_get_events();
-				tablet_previously_succeeded = true;
-			} else {
-				// Failure in tablet control.  This would be better if done
-				// at the selection stage before the tablet is even used.
-				failsafe_control_select(reset_controller || tablet_previously_succeeded);
-				control_events = failsafe_control_get_events();
-				tablet_previously_succeeded = false;
-			}
+			ret = tablet_control_select(reset_controller);
 			break;
-		}
 		case FLIGHTSTATUS_CONTROLSOURCE_GEOFENCE:
-			geofence_control_select(reset_controller);
-			control_events = geofence_control_get_events();
+			ret = geofence_control_select(reset_controller);
 			break;
-		case FLIGHTSTATUS_CONTROLSOURCE_FAILSAFE:
 		default:
-			failsafe_control_select(reset_controller);
-			control_events = failsafe_control_get_events();
+			/* Fall through into failsafe */
 			break;
 		}
+
+		if (ret) {
+			failsafe_control_select(last_control_selection !=
+					FLIGHTSTATUS_CONTROLSOURCE_FAILSAFE);
+			control_selection = FLIGHTSTATUS_CONTROLSOURCE_FAILSAFE;
+		}
+
 		if (control_selection != last_control_selection) {
 			FlightStatusControlSourceSet(&control_selection);
 			last_control_selection = control_selection;
 		}
 
-		// TODO: This can evolve into a full FSM like I2C possibly
-		switch(control_events) {
-		case CONTROL_EVENTS_NONE:
-			break;
-		case CONTROL_EVENTS_ARM:
-			control_event_arm();
-			break;
-		case CONTROL_EVENTS_ARMING:
-			control_event_arming();
-			break;
-		case CONTROL_EVENTS_DISARM:
-			control_event_disarm();
-			break;
+		uint32_t now = PIOS_Thread_Systime();
+#define GO_STATE(x) \
+				do { \
+					arm_state = (x); \
+					arm_state_time = (now); \
+				} while (0)
+
+		/* Global state transitions: weird stuff means disarmed +
+		 * safety hold-down.
+		 */
+		if ((status == STATUS_ERROR) ||
+				(status == STATUS_SAFETYTIMEOUT)) {
+			GO_STATE(ARM_STATE_SAFETY);
 		}
 
-		// Wait until next update
-		PIOS_RCVR_WaitActivity(UPDATE_PERIOD_MS);
-		PIOS_WDG_UpdateFlag(PIOS_WDG_MANUAL);
-	}
-}
+		/* If there's an alarm that prevents arming, enter a 0.2s
+		 * no-arming-holddown.  This also stretches transient alarms
+		 * and makes them more visible.
+		 */
+		if (!ok_to_arm() && (arm_state != ARM_STATE_ARMED)) {
+			GO_STATE(ARM_STATE_SAFETY);
+		}
 
-//! When the control system requests to arm the FC
-static int32_t control_event_arm()
-{
-	if(ok_to_arm()) {
-		FlightStatusData flightStatus;
-		FlightStatusGet(&flightStatus);
-		if (flightStatus.Armed != FLIGHTSTATUS_ARMED_ARMED) {
-			flightStatus.Armed = FLIGHTSTATUS_ARMED_ARMED;
-			FlightStatusSet(&flightStatus);
-			vehicle_is_armed = true;
+		switch (arm_state) {
+			case ARM_STATE_SAFETY:
+				/* state_time > 0.2s + "NONE" or "DISARM" -> DISARMED
+				 *
+				 * !"NONE" and !"DISARM" -> SAFETY, reset state_time
+				 */
+				if ((status == STATUS_NONE) ||
+						(status == STATUS_DISARM)) {
+					if ((now - arm_state_time) > 200) {
```

Comment:
This is making sure that things are "normal" for a bit (disarmed, no alarms) before allowing an arming event.  It makes sure if an RX is freaking out that it still needs to present as normally connected, disarmed for awhile before the arming event can happen, and that transient severe alarms prevent arming a greater portion of the time.